rfcs/taskforce/0003-Integration-test-infrastructure-spec.md

```diff
@@ -0,0 +1,204 @@
+    ; archive_nodes: archive_nodes }
+
+```
+
+3.Integration test: This is the program that defines the integration test written in a custom DSL (#4766)
+The inputs to the test executive (network config and runtime config) can be passed to this program
+
+A test executive will interact with the two main components as shown in the image :
+
+1. Network manager
+2. Log engine
+
+<img src="../res/test_exec.png" alt="drawing" width="500"/>
+
+##### Network Manager
+
+A network manager is basically a wrapper around the network infrastructure using which a testnet is deployed. The network manager implemented in ocaml would take as input a network configuration and use it to generate required configuration for the deployment and deploy a testnet. The network manager is responsible for creating the seed nodes that other nodes can connect to. A separate network manager would be created for each environment where we want to run the testnet. The choice of a network manager is made by the test executive based on a user input (whether to run locally or on the cloud). The interface between a network manager and the test executive exposes all the nodes and the commands that can be issued to the nodes.
+
+In the current testnet infrastructure, a network manager would generate the necessary terraform configuration files and deploy using terraform.
+To run it locally, using minikube perhaps, we would create another network manager that deploys a testnet locally.
+
+The network manager itself can be specified as:
+
+```ocaml
+
+module Node : Node_intf
+(* Deploy the network*)
+val deploy : Network_config.t -> Daemon_config.t -> Testnet.t Deferred.t
+
+(*Tear down the network*)
+val destroy : Testnet.t -> unit Deferred.t
+```
+
+```ocaml
+
+module type Node_intf : sig
+    type t
+
+    val node_id_json : t -> json
+
+    val start : t -> unit Deferred.t
+
+    val stop : t -> unit Deferred.t
+
+    ...
+end
+
+module type Testnet = sig
+
+    type t =
+        {
+          block_producers : Node.t list
+        ; snark_coordinators: Node.t list
+        ; archive_nodes: Node.t list
+        ; ... }
+end
+```
+
+A node will encapsulate all the information needed to connect to it and make requests such as sending payments, retrieving a node's status, and so on. These requests implemented by the network manager could be either using graphql commands or cli commands. The interface should also implement functions to start and stop a node that can be triggered at any time by the test program. This is particularly needed for testing functionalities like bootstrap and catchup. The network infrastructure should facilitate stopping and restarting coda daemon containers (with or w/o persisting the config directory). Currently a pod or a container within a pod is restarted automatically every time the daemon terminates. This behavior should be changed to support restarting a node on request. The following are some of the options to enable restart on request:
```

Comment:
I think consensus previously was that CLI commands should all be backed by GraphQL commands. It would be good to commit here to only using GraphQL, and add GraphQL implementations to CLI commands missing them if we need them.